internals/daemon/access.go

```diff
@@ -0,0 +1,60 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Based on: https://github.com/snapcore/snapd/blob/master/daemon/access.go
+
+package daemon
+
+import (
+	"net/http"
+)
+
+// accessChecker checks whether a particular request is allowed.
+//
+// An access checker will either allow a request (returns nil) or deny it.
+type accessChecker interface {
+	CheckAccess(d *Daemon, r *http.Request, ucred *ucrednet, user *UserState) Response
+}
+
+// openAccess allows all requests
+type openAccess struct{}
+
+func (ac openAccess) CheckAccess(d *Daemon, r *http.Request, ucred *ucrednet, user *UserState) Response {
+	return nil
+}
+
+// rootAccess allows requests from the root uid
+type rootAccess struct{}
+
+func (ac rootAccess) CheckAccess(d *Daemon, r *http.Request, ucred *ucrednet, user *UserState) Response {
+	if ucred != nil && ucred.Uid == 0 {
```

Comment:
In case of snapd, there are two sockets (introduced in https://github.com/snapcore/snapd/pull/1749):

```golang
SnapdSocket = filepath.Join(rootdir, "/run/snapd.socket")
SnapSocket = filepath.Join(rootdir, "/run/snapd-snap.socket")
```

It seems like `snapd.socket` is accessed from the `snap` command and with privileged commands and `snapd-snap.socket` is accessed from `snapctl` and from within (possibly confined) snaps, and is the "public" API.

As Pebble doesn't have the concept of two different sockets with split API yet (it only listens on the UNIX domain socket + optionally on a TCP socket, and only the UNIX domain socket has `ucred != nil`), I don't think there's a good way to "port" this over to Pebble for now.

Reply:
I just saw when looking at the tests that `Daemon` actually has a `untrustedSocketPath` and `untrustedListener`, is that something that's actively used in Pebble? (`untrustedSocketPath: opts.SocketPath + ".untrusted"`)

Reply:
Well, seems like `UntrustedOK` was used as the switch for allowing access from the `untrustedSocketPath`, but `UntrustedOK` is nowhere used at the moment (except in the now-obsolete `canAccess()` and the unit tests).

Reply:
I could also split removal of `UntrustedOK` into a separate PR before this one is merged, so that it's more obvious that it isn't used, and it's clearer what gets removed.

Reply:
This is now #361 and removes also the socket itself.